### PR TITLE
WIP Add eclfio_array_get/put

### PR DESCRIPTION
**Pull request by jokva**
These functions provide an abstraction for the eclipse detail of large
arrays being written across multiple smaller physical blocks.

In effect, it enables reading an eclipse record (the header
keyword-size-type, followed by the array) as:

    header = eclfio_get()
    nmemb = parse-header(header)
    return eclfio_array_get(nmemb)

Remaining work:

- tests!
- some considerations on return types
- port ecl2 to use this

I also propose we add a new data type in opts, `C` or `s`, for the fixed len-8 string. That means the `len` argument to the array function really only makes sense for var-strings (which are special) funky cool functionality where you emulate 64-bit ints by merging two consecutive int32. 

**Task**
_Short description of the task_

**Approach**
_Short description of the approach_

**Pre un-WIP checklist**
- [ ] Statoil tests pass locally

### CODE REVIEW

**Comment by joakim-hove** [comment]
> also propose we add a new data type in opts, C or s, for the fixed len-8 string.

We could even parse the type Cnnn - then we do not have to distingush between the 8 charatcter type and variable length type.

File: lib/src/fortio.cpp [hunk 1]
```diff
@@ -409,3 +409,67 @@ int eclfio_put( std::FILE* fp,
...
+
+int eclfio_array_get( std::FILE* fp,
+                      const char* opts,
+                      int len,
+                      int nmemb,
+                      void* array ) {
+
+    if( nmemb < 0 ) return ECL_EINVAL;
+
+    if( len < 1 ) len = 1;
```
Comment (joakim-hove) [inline lib/src/fortio.cpp]: What is the logic here?

**Comment by jokva** [comment]
> We could even parse the type Cnnn - then we do not have to distingush between the 8 charatcter type and variable length type.

Surely, but on the flipside that *greatly* complicates the parseopts function.

File: lib/src/fortio.cpp (hunk 1 above)
Comment (jokva) [reply lib/src/fortio.cpp]: Just robustness, allowing zero or negative values as "don't worry about len at all), to avoid nasty divide-by-zeros.

**Comment by joakim-hove** [comment]
> Surely, but on the flipside that greatly complicates the parseopts function.

That is obviously true - but *iff* that leads to a unified treatment of the CHAR and Cnnn types it might be worth it?

**Comment by jokva** [comment]
> That is obviously true - but iff that leads to a unified treatment of the CHAR and Cnnn types it might be worth it?

Maybe, but having a concatenation feature has some pretty cool implications too, because then you can `put` your record headers as:

`eclfio_array_put(fp, "c", 16, 1, 1000, header)`

Which serves nicely as documentaiton.

**Comment by jokva** [comment]
> `eclfio_array_put(fp, "c", 16, 1, 1000, header)`

In this context, adding `b` as an alias (`byte`) for `c` could also be useful.

File: lib/e3/ecl/fortio.h [hunk 2]
```diff
@@ -166,6 +166,70 @@ int eclfio_get( FILE*, const char* opts, int32_t* recordsize, void* record );
...
+ */
+#define ECL_DEFAULT_BLOCKSIZE_NUMERIC 1000
+#define ECL_DEFAULT_BLOCKSIZE_STRING  105
+int ecl_default_blocksize( int );
+
+/*
+ * eclfio_array_get/put behave differently w.r.t. errors than their simpler
+ * relatives eclfio_get/put.
+ *
+ * When an operation cannot be successfully completed, eclfio_array does *not*
```
Comment (joakim-hove) [inline lib/e3/ecl/fortio.h]: I think these functions will be the main atomic units; so personally I think argument parsing, roll-back guarantees and so on should be at this level. The fortio level is not interesting for the user,

File: lib/e3/ecl/fortio.h [hunk 3]
```diff
@@ -166,6 +166,70 @@ int eclfio_get( FILE*, const char* opts, int32_t* recordsize, void* record );
...
+
+int eclfio_array_get( FILE*,
+                      const char* opts,
+                      int len,
+                      int nmemb,
+                      void* array );
+int eclfio_array_put( FILE*,
+                      const char* opts,
+                      int len,
+                      int size,
```
Comment (joakim-hove) [inline lib/e3/ecl/fortio.h]: Extra size in header.

**Comment by joakim-hove** [comment]
> In this context, adding b as an alias (byte) for c could also be useful.

Agree

**Comment by joakim-hove** [comment]
This look fine to me; two things:

1. This is the time to get the  CHAR / Cnnnn implementation right. Maybe also the bool?
2. Personally I think these functions should be fundamental user exposed functions, .i.e. options parsing and rollback guard should be handled from here - the one level deeper functions from the previous PR should get a already parsed options object - and rollback should be handled here,

File: lib/e3/ecl/fortio.h (hunk 2 above)
Comment (jokva) [reply lib/e3/ecl/fortio.h]: > I think these functions will be the main atomic units; so personally I think argument parsing, roll-back guarantees and so on should be at this level. The fortio level is not interesting for the user,

Who's _the user_?

I have mixed feelings about this. If rollback is done on array, then we'd either drop or have to communicate _which_ block failed, but if it only rolls back to the point that failed, then there's immediate recoverability.

Also, that means reading record headers can read with `eclfio_get` directly, with rollback semantics, which is nice.

That being said, `array_get` *will* roll back if nmemb < size.

Comment (joakim-hove) [reply lib/e3/ecl/fortio.h]: > Who's the user?

The "user" in this case is just code - probably same library, slighter higher up in the stack.

> If rollback is done on array, then we'd either drop or have to communicate which block failed, but if it only rolls back to the point that failed, then there's immediate recoverability.

Yes - my point is that array is the relevant atomic unit - in all(?) contexts; i.e. I do not see it as interesting functionality to roll back to middle of an array - i.e. a record.

> Also, that means reading record headers can read with eclfio_get directly, with rollback semantics, which is nice.

I agree that reading headers with rollback semantics is nice - and important, but they can - and should(?) - be implemented as array_get of 16 characters - i.e. again array smenatics; although the array in this case always only consists of one block.

Comment (jokva) [reply lib/e3/ecl/fortio.h]: > The "user" in this case is just code - probably same library, slighter higher up in the stack.

Sure, which is why I personally like this. It allows to check that specific block with sizeof, and quickly determine that the fortran block was of the wrong size, and maybe even pad the output array accordingly for robustness.

> Yes - my point is that array is the relevant atomic unit - in all(?) contexts; i.e. I do not see it as interesting functionality to roll back to middle of an array - i.e. a record.

The main use case is error handling, and I think the cost in interface is worth it personally.

**Comment by markusdregi** [comment]
@jokva I don't see this work moving anytime soon, so please make a note as you see fit (mental, GitHub issue or as a Post-it note on your keyboard) of the potential gold in this work. I'm closing the PR until this is continued. Please reopen if you disagree.

**Pull request closed without merging**